// @replayio/cypress/dist/features.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isFeatureEnabled = exports.getFeatures = exports.PluginFeature = void 0;
var PluginFeature;
(function (PluginFeature) {
    PluginFeature["Plugin"] = "plugin";
    PluginFeature["Support"] = "support";
    PluginFeature["Metrics"] = "metrics";
})(PluginFeature || (exports.PluginFeature = PluginFeature = {}));
function isValidFeatureOption(feature) {
    if (["all", "none"].includes(feature)) {
        return true;
    }
    if (feature.startsWith("no-")) {
        feature = feature.substring(3);
    }
    return Object.values(PluginFeature).includes(feature);
}
function parsePluginFeatureOptions(options) {
    return options
        .toLowerCase()
        .split(",")
        .map(s => s.trim())
        .filter(s => isValidFeatureOption(s));
}
function getFeatures(options) {
    const allFeatures = Object.values(PluginFeature);
    if (options) {
        return parsePluginFeatureOptions(options).reduce((acc, feature) => {
            if (feature === "all") {
                return allFeatures;
            }
            else if (feature === "none") {
                return [];
            }
            else if (feature.startsWith("no-")) {
                feature = feature.substring(3);
                if (acc.includes(feature)) {
                    return acc.filter(f => f !== feature);
                }
            }
            else if (!acc.includes(feature)) {
                acc.push(feature);
            }
            return acc;
        }, []);
    }
    return allFeatures;
}
exports.getFeatures = getFeatures;
function isFeatureEnabled(options, feature) {
    const features = getFeatures(options);
    return features.includes(feature);
}
exports.isFeatureEnabled = isFeatureEnabled;
