// @replayio/test-utils/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initMetadataFile = exports.getMetadataFilePath = exports.ReplayReporter = exports.fetchWorkspaceConfig = exports.removeAnsiCodes = exports.pingTestMetrics = exports.ReporterError = exports.buildTestId = void 0;
const reporter_1 = __importDefault(require("./reporter"));
exports.ReplayReporter = reporter_1.default;
var testId_1 = require("./testId");
Object.defineProperty(exports, "buildTestId", { enumerable: true, get: function () { return testId_1.buildTestId; } });
var reporter_2 = require("./reporter");
Object.defineProperty(exports, "ReporterError", { enumerable: true, get: function () { return reporter_2.ReporterError; } });
var metrics_1 = require("./metrics");
Object.defineProperty(exports, "pingTestMetrics", { enumerable: true, get: function () { return metrics_1.pingTestMetrics; } });
var terminal_1 = require("./terminal");
Object.defineProperty(exports, "removeAnsiCodes", { enumerable: true, get: function () { return terminal_1.removeAnsiCodes; } });
var config_1 = require("./config");
Object.defineProperty(exports, "fetchWorkspaceConfig", { enumerable: true, get: function () { return config_1.fetchWorkspaceConfig; } });
__exportStar(require("./logging"), exports);
var metadata_1 = require("./metadata");
Object.defineProperty(exports, "getMetadataFilePath", { enumerable: true, get: function () { return metadata_1.getMetadataFilePath; } });
Object.defineProperty(exports, "initMetadataFile", { enumerable: true, get: function () { return metadata_1.initMetadataFile; } });
